Share Annotations with inner resources in Azure PostgreSQL, Redis, and SQL Server resources.

eerhardt opened the pull request:
This allows for things like WaitFor to work correctly when using RunAsContainer.

Fix #6146
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/6164)

eerhardt committed:
Share Annotations with inner resources in Azure PostgreSQL, Redis, and SQL Server resources.

This allows for things like WaitFor to work correctly when using RunAsContainer.

Fix #6146

eerhardt commented:
/backport to release/9.0-rc1

mitchdenny commented on src/Aspire.Hosting.Azure.PostgreSQL/AzurePostgresExtensions.cs:
```diff
@@ -289,6 +289,11 @@ public static IResourceBuilder<AzurePostgresFlexibleServerResource> RunAsContain
+            .Where(db => db.Parent == azureResource)
+            .ToDictionary(db => db.Name);
+
         RemoveAzureResources(builder.ApplicationBuilder, azureResource);
```
Not a big issue, but this results in scanning the resource collection twice to get the same set of data from above. Why not just iterate over the resources inline or do the deletes in the loop below.

eerhardt replied:
I can pass the dictionary into the Remove method.

mitchdenny commented on tests/Aspire.Hosting.Azure.Tests/AzurePostgresExtensionsTests.cs:
```diff
@@ -323,4 +323,52 @@ public async Task WithPasswordAuthenticationBeforeAfterRunAsContainer(bool befor
+    [Theory]
+    [InlineData(true)]
+    [InlineData(false)]
+    public void RunAsContainerAppliesAnnotationsCorrectly(bool before)
```
What happens if I do this?

```csharp
var db = builder.AddAzurePostgresFlexibleServer("pg")
                .WithPasswordAuthentication()
                .RunAsContainer(c =>
                {
                    c.WithPgAdmin();
                })
                .AddDatabase("db")
                .WithAnnotation(new DummyAnnotation());
```

Based on what I'm seeing in this branch the DummyAnnotation will be lost by the time this gets down into the application executor.

eerhardt replied:
Good catch! I've fix this and added tests for it.

mitchdenny commented:
I just realized that you are going to need to update `AddDatabase(...)` as well to copy annotations if the parent resource has an InnerResource.

https://github.com/dotnet/aspire/blob/770b3dfeeab32afc5403d4cee5131be723faf2c0/src/Aspire.Hosting.Azure.PostgreSQL/AzurePostgresExtensions.cs#L256

eerhardt committed:
Address PR feedback

- Reuse dictionary of databases to remove them from the model
- Fix a bug when AddDatabase is called after RunAsContainer

eerhardt commented:
> I just realized that you are going to need to update `AddDatabase(...)` as well to copy annotations if the parent resource has an InnerResource.
> 
> https://github.com/dotnet/aspire/blob/770b3dfeeab32afc5403d4cee5131be723faf2c0/src/Aspire.Hosting.Azure.PostgreSQL/AzurePostgresExtensions.cs#L256

Yep, that was the fix for https://github.com/dotnet/aspire/pull/6164#discussion_r1791031810.

mitchdenny reviewed (approved):
LGTM ... verified that commands annotations carry across.
